client/src/modules/authorization/components/SignUpForm.jsx

```diff
@@ -0,0 +1,319 @@
+    this.setState(
+      { passwordConfirm, isPasswordConfirmValid: isValid },
+      this.comparePasswords
+    );
+  };
+
+  nameValidator = value => {
+    const emptyInfoId = 'authorization.input.username.empty';
+    const errorInfoId = 'authorization.input.username.invalid';
+    const { isEmpty, isLength } = validator;
+
+    if (isEmpty(value)) {
+      return emptyInfoId;
+    }
+
+    if (!isLength(value, { min: 1, max: 32 })) {
+      return errorInfoId;
+    }
+
+    return '';
+  };
+
+  emailValidator = value => {
+    const emptyInfoId = 'authorization.input.email.empty';
+    const errorInfoId = 'authorization.input.email.invalid';
+    const { isEmpty, isEmail } = validator;
+
+    if (isEmpty(value)) {
+      return emptyInfoId;
+    }
+
+    if (!isEmail(value)) {
+      return errorInfoId;
+    }
+
+    return '';
+  };
+
+  passwordValidator = value => {
+    const emptyInfoId = 'authorization.input.password.empty';
+    const errorInfoId = 'authorization.input.password.invalid';
+    const { isEmpty, isLength } = validator;
+
+    if (isEmpty(value)) {
+      return emptyInfoId;
+    }
+
+    if (!isLength(value, { min: 4 })) {
+      return errorInfoId;
+    }
+
+    return '';
+  };
+
+  comparePasswords = () => {
+    const {
+      errors,
+      errors: { passwordError },
+      password,
+      passwordConfirm
+    } = this.state;
+    let newError;
+
+    if (passwordError && password === passwordConfirm) {
+      newError = '';
+    }
+
+    if (
+      !passwordError &&
+      password !== '' &&
+      passwordConfirm !== '' &&
+      password !== passwordConfirm
+    ) {
+      newError = 'authorization.input.password.not-same';
+    }
+
+    if (newError !== undefined) {
+      this.setState({
+        errors: {
+          ...errors,
+          passwordError: newError
+        }
+      });
+    }
+  };
+
+  isFormValid = () => {
+    const {
+      errors,
+      isEmailValid,
+      isNameValid,
+      isPasswordConfirmValid,
+      isPasswordValid
+    } = this.state;
+
+    const isError = _find(errors, error => error !== '');
+
+    if (
+      isNameValid &&
+      isEmailValid &&
+      isPasswordValid &&
+      isPasswordConfirmValid &&
+      !isError
+    ) {
+      return this.setState({ isFormValid: true });
+    }
+
+    this.setState({ isFormValid: false });
+  };
+
+  handleSignUp = () => {
+    const { signUp } = this.props;
+    const { email, name, password, passwordConfirm } = this.state;
+
+    this.setState({ pending: true });
+
+    this.pendingPromise = makeAbortablePromise(
+      signUp(email, name, password, passwordConfirm)
+    );
+    this.pendingPromise.promise
+      .then(() => this.setState({ confirmationSend: true, pending: false }))
+      .catch(err => {
+        if (!(err instanceof AbortPromiseException)) {
+          const { status } = err.response;
+          const newState = { pending: false };
+
+          if (status === 406) {
```

Comment:
why checking status here?

Reply:
I thought that we will need some backend validation for this sign-up form (according to the rule: "never trust frontend"). I'm not sure if this is really needed but If so, and we will check form fields on the backend again and backend will find errors we will have to inform the user about these errors - send appropriate status (406) and error messages to display them on the frontend. I tried to prepare this method for this case and for status 406 react validation errors sent by the backend. We try to build the app without returning messages from the backend but is there any other way to validate the form on the backend without returning specific status and messages which inform the user what's went wrong in specific fields?

Reply:
But why do that here, why not in action? The action can also throw with `errors` object. The last place where the status should be checked is in actions.